assets/src/components/routeLadder.tsx

```diff
@@ -0,0 +1,59 @@
+import React, { useEffect } from "react"
+import { fetchTimepointsForRoute } from "../api"
+import { Route, Timepoint, TimepointsByRoute } from "../skate"
+import { Dispatch, setTimepointsForRoute } from "../state"
+import Loading from "./loading"
+
+interface Props {
+  route: Route
+  dispatch: Dispatch
+  timepointsByRoute: TimepointsByRoute
+}
+
+const TimepointStop = () => (
+  <div className="m-route-ladder__timepoint-stop">
+    <svg height="60" width="20">
+      <line x1="10" y1="0" x2="10" y2="25" />
+      <circle cx="10" cy="30" r="5" />
+      <line x1="10" y1="35" x2="10" y2="60" />
+    </svg>
+  </div>
+)
+
+const Timepoint = ({ timepoint }: { timepoint: Timepoint }) => (
+  <li className="m-route-ladder__timepoint">
+    <TimepointStop />
+    <div className="m-route-ladder__timepoint-name">{timepoint.id}</div>
+    <TimepointStop />
+  </li>
+)
+
+const RouteLadder = ({ route, dispatch, timepointsByRoute }: Props) => {
+  const timepoints = timepointsByRoute[route.id]
+
+  useEffect(() => {
+    if (!timepointsByRoute[route.id]) {
+      fetchTimepointsForRoute(route.id).then((newTimepoints: Timepoint[]) =>
+        dispatch(setTimepointsForRoute(route.id, newTimepoints))
+      )
+    }
+  })
```

Comment:
This effect needs something to make sure it's not called multiple times, since it could be rendered multiple times before we get the response.

Alternatively, maybe this should live in the reducer, in the case for `selectRoute`. That's how Elm does it: the reducer both updates the state and initiates any new side effects. I don't know if the React convention is to put side effects in the components or the reducer.

Reply:
Good call. I added a null state for loading and updated the check to only care about `undefined`.

Hooks need to be in React components so we can't move it to the reducer, but we should keep in mind the ability to create custom hooks if we need to share or abstract them in the future. (I don't think there's value in doing that here at this point.)